src/common/hashes/shamap.ts

```diff
@@ -0,0 +1,288 @@
+import hashPrefix from './hash-prefix'
+import hash from './sha512hash'
+const HEX_ZERO = '0000000000000000000000000000000000000000000000000000000000000000'
+
+export enum NodeTypes {
```

Comment:
nit: generally enums are named as a singular, `NodeType` vs `NodeTypes`.

This makes your code below more readable
```
// Sounds like this should be an array?  
public type: NodeTypes
```

changes to 
```
  // Ah, a singular enum value
  public type: NodeType
```